src/Language/Haskell/TH/Datatype.hs

```diff
@@ -316,37 +317,120 @@ normalizeDec' context name params cons variant =
+                  RecC n xs ->
+                    let fns = takeFieldNames xs in
+                    pure [ConstructorInfo n tyvars context
+                            (takeFieldTypes xs) (RecordConstructor fns)]
+                  ForallC tyvars' context' c' ->
+                    go (tyvars'++tyvars) (context'++context) c'
 #if MIN_VERSION_template_haskell(2,11,0)
-        GadtC ns xs innerType ->
-          gadtCase ns innerType (map snd xs) NormalConstructor
-        RecGadtC ns xs innerType ->
-          let fns = takeFieldNames xs in
-          gadtCase ns innerType (takeFieldTypes xs) (RecordConstructor fns)
-      where
-        gadtCase = normalizeGadtC typename params tyvars context
-
+                  GadtC ns xs innerType ->
+                    gadtCase ns innerType (map snd xs) NormalConstructor
+                  RecGadtC ns xs innerType ->
+                    let fns = takeFieldNames xs in
+                    gadtCase ns innerType (takeFieldTypes xs) (RecordConstructor fns)
+                where
+                  gadtCase = normalizeGadtC typename params tyvars context
+#endif
+#if MIN_VERSION_template_haskell(2,8,0) && (!MIN_VERSION_template_haskell(2,10,0))
+          dataFamCompatCase :: Con -> Q [ConstructorInfo]
+          dataFamCompatCase = go []
+			where
+              go tyvars c =
+                case c of
+                  NormalC n _ ->
+                    dataFamCase' n tyvars NormalConstructor
+                  InfixC _ n _ ->
+                    dataFamCase' n tyvars NormalConstructor
+                  RecC n xs ->
+                    dataFamCase' n tyvars (RecordConstructor (takeFieldNames xs))
+                  ForallC tyvars' context' c' ->
+                    go (tyvars'++tyvars) c'
+
+          dataFamCase' :: Name -> [TyVarBndr] -> ConstructorVariant
+                       -> Q [ConstructorInfo]
+          dataFamCase' n tyvars variant = do
+            info <- reify n
+            case info of
+              DataConI _ ty _ _ -> do
+                let (context, argTys :|- returnTy) = uncurryType ty
+                returnTy' <- resolveTypeSynonyms returnTy
+                -- Notice that we've ignored the Cxt and argument Types from the
+                -- Con argument above, as they might be scoped over eta-reduced
+                -- variables. Instead of trying to figure out what the
+                -- eta-reduced variables should be substituted with post facto,
+                -- we opt for the simpler approach of using the context and
+                -- argument types from the reified constructor Info, which will
+                -- at least be correctly scoped. This will make the task of
+                -- substituting those types with the variables we put in
+                -- place of the eta-reduced variables (in normalizeDec)
+                -- much easier.
+                normalizeGadtC typename params tyvars context [n]
```

Comment:
This `normalizeGadtC` function came in handy here! However, it might be a bit inappropriately named now, since here it's being used on things that aren't necessarily GADTs.